client/state/user-settings/actions.js

```diff
@@ -0,0 +1,171 @@
+/**
+ * External dependencies
+ */
+import debugFactory from 'debug';
+import { isEmpty, keys } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+import userUtils from 'lib/user/utils';
+import {
+	USER_SETTINGS_FETCH,
+	USER_SETTINGS_FETCH_SUCCESS,
+	USER_SETTINGS_FETCH_FAILURE,
+	USER_SETTINGS_SAVE_FAILURE,
+	USER_SETTINGS_SAVE_SUCCESS,
+	USER_SETTINGS_RECEIVE,
+	USER_SETTINGS_UNSAVED_SET,
+	USER_SETTINGS_UNSAVED_REMOVE,
+} from 'state/action-types';
+
+const debug = debugFactory( 'calypso:user:settings' );
+
+/**
+ * Fetch user settings from WordPress.com API and store them in UserSettings instance
+ * @returns {Function} Action thunk
+ */
+export function fetchUserSettings() {
+	return ( dispatch, getState ) => {
+		if ( ! userUtils.isLoggedIn() || getState().userSettings.fetchingSettings ) {
+			return;
+		}
+
+		dispatch( { type: USER_SETTINGS_FETCH } );
+
+		debug( 'Fetching user settings' );
+
+		const fetchRequest = wpcom.me().settings().get();
+
+		fetchRequest.then( ( data ) => {
+			dispatch( receiveUserSettings( data ) );
+			dispatch( { type: USER_SETTINGS_FETCH_SUCCESS } );
+
+			debug( 'Settings successfully retrieved' );
+		} ).catch( ( error ) => {
+			dispatch( {
+				type: USER_SETTINGS_FETCH_FAILURE,
+				error
+			} );
+		} );
+
+		// Let the caller know whether the fetch was successful or not
+		return fetchRequest;
+	};
+}
+
+/**
+ * Post settings to WordPress.com API at /me/settings endpoint
+ *
+ * @param {Object} settingsOverride - default settings object
+ * @return {Function} Action thunk
+ */
+export function saveUserSettings( settingsOverride ) {
+	return ( dispatch, getState ) => {
+		const settings = settingsOverride ? settingsOverride : getState().userSettings.unsavedSettings;
+
+		if ( isEmpty( settings ) ) {
+			debug( 'There are no settings to save.' );
+			return Promise.resolve( null );
+		}
+
+		debug( 'Saving settings: ' + JSON.stringify( settings ) );
+
+		const saveRequest = wpcom.me().settings().update( settings );
+
+		saveRequest.then( ( data ) => {
+			dispatch( receiveUserSettings( data ) );
+			dispatch( {
+				type: USER_SETTINGS_SAVE_SUCCESS,
+				savedSettingNames: settingsOverride ? keys( settingsOverride ) : null
+			} );
+
+			// Refetch the user data after saving user settings
+			// user.fetch();
+		} ).catch( ( error ) => {
+			dispatch( {
+				type: USER_SETTINGS_SAVE_FAILURE,
+				error
+			} );
+		} );
+
+		// Let the caller know whether the save was successful or not
+		return saveRequest;
+	};
+}
+
+/**
+ * Returns an action object signalling the settings have been received from server.
+ *
+ * @param  {Object} values Settings values
+ * @return {Object}        Action object
+ */
+export function receiveUserSettings( values ) {
+	return {
+		type: USER_SETTINGS_RECEIVE,
+		values
```

Comment:
notice here how we have hidden any clues about the shape and structure of `values`

I hoped that by looking into the action creator I could discover what is expected there but it left me without much help. what can we do in cases like this? well we could provide some examples in the JSDoc of what this could look like, describe in the `@param` tag, or explicitly show the shape by using more split-up values in the action itself